### ISSUE TITLE
ZFS Volume Size Reporting

### ISSUE THREAD

**Issue by NIronwolf** [NebN/unraid-simple-monitoring-api#9]
Any volume (or my 4 disk cache pool) that is formatted under zfs always reports 0 bytes used.  This also throws off the percentages.  Are there any plans to support these volumes better?  It also affects other monitoring tools and the unraid dashboard seems to be the only place that reports the sizes correctly.

Thank you!

**Comment by NebN** [comment]
That's curious. I have an old disk lying around I could format as zfs to try and figure out the problem. 

In the meantime it would be very helpful if you could go to the application's logs and see if there are any errors being logged.

**Comment by NIronwolf** [comment]
There were no errors in the logs.  I think it's just how zfs reports to normal OS functions.

Here is what zfs list and df -h report for the sizes on my zfs formated volumes and what the web GUI shows.

(nvme_cache is 4x 2TB in raidz1 so the ~6TB size is expected.)

```
zfs list
NAME            USED  AVAIL     REFER  MOUNTPOINT
disk1          12.1T  5.98T      144K  /mnt/disk1
disk10         8.99T  9.07T      104K  /mnt/disk10
disk11         8.97T  9.09T      104K  /mnt/disk11
disk6          13.6T  4.42T      104K  /mnt/disk6
nvme_cache     4.78T   508G      209K  /mnt/nvme_cache
ssd_cache      59.3G  1.70T     6.58G  /mnt/ssd_cache
```

```
df -h
Filesystem     Size  Used Avail Use% Mounted on
disk1          6.0T  256K  6.0T   1% /mnt/disk1
disk6          4.5T  128K  4.5T   1% /mnt/disk6
disk10         9.1T  128K  9.1T   1% /mnt/disk10
disk11         9.1T  128K  9.1T   1% /mnt/disk11
nvme_cache     508G  256K  508G   1% /mnt/nvme_cache
ssd_cache      1.8T  6.6G  1.7T   1% /mnt/ssd_cache
```

![image](https://github.com/NebN/unraid-simple-monitoring-api/assets/635345/49e03549-6abc-4e19-af10-0eecc8056588)

**Comment by NebN** [comment]
Thank you for the details. I'll see what I can figure out.

**Comment by NebN** [comment]
I am unfortunately unable to reproduce this bug with a single ZFS disk. As you can see I have a single cache disk formatted as ZFS.
![image](https://github.com/NebN/unraid-simple-monitoring-api/assets/57036949/56b6d2db-89f9-471e-871f-79b8a8a65d42)

This is my conf.yml
```yaml
disks:
  cache:
    - /mnt/cache
    - /mnt/test
```

And this is the response I get
```json
"cache":[{"mount":"/mnt/cache","total":465,"used":327,"free":138,"used_percent":70.32,"free_percent":29.68},{"mount":"/mnt/test","total":430,"used":4,"free":426,"used_percent":0.93,"free_percent":99.07}]
```

I suppose the problem only appears with raidzN configurations, I do happen to have 3 random disks I could try and cobble together into a ZFS raidz1 pool. I'll see what I can do.

**Comment by NebN** [comment]
> Are the files on it _in_ a dataset?

This seems to be the issue. Forgive me but I am not very knowledgeable when it comes with ZFS. I currently have a raidz1 configuration with files in a dataset, and the API is reporting disk usage as 0.

_However_  I was able to fix this by specifying the dataset mount, as opposed to the cache mount itself.

Conf
```yaml
disks:
  cache:
    - /mnt/cache
    - /mnt/cache_test
    - /mnt/cache_test/test_dataset
```

This is the response, cache_test is empty, but cache_test/test_dataset is correctly being read
```json
{"mount":"/mnt/cache","total":465,"used":327,"free":138,"used_percent":70.32,"free_percent":29.68},
{"mount":"/mnt/cache_test","total":211,"used":0,"free":211,"used_percent":0,"free_percent":100},
{"mount":"/mnt/cache_test/test_dataset","total":215,"used":4,"free":211,"used_percent":1.86,"free_percent":98.14}
```
You'd want to remove /mnt/cache_test from the conf, to obtain the correct total amount.

_However however_, with  multiple datasets this doesn't work as they are seen as different disks and their sizes are added.

```json
{"mount":"/mnt/cache","total":465,"used":327,"free":138,"used_percent":70.32,"free_percent":29.68},
{"mount":"/mnt/cache_test","total":203,"used":0,"free":203,"used_percent":0,"free_percent":100},
{"mount":"/mnt/cache_test/test_dataset","total":207,"used":4,"free":203,"used_percent":1.93,"free_percent":98.07},
{"mount":"/mnt/cache_test/test_dataset_2","total":211,"used":8,"free":203,"used_percent":3.79,"free_percent":96.21}
```

I see two solutions:
- best and more difficult: figure out the root cause of the issue, and make it so the software is able to tell that the specified mount is a ZFS pool and extract the correct values
- quick and dirty: add a new configuration section specifically for ZFS datasets to be able to interpret the data correctly when reading the various datasets

I'll attempt the first solution and fallback to the second if it's not feasible.

**Comment by NebN** [comment]
I've been looking into possible solutions. However I'm missing some crucial information, namely I am not sure I understand the interaction between the 4 ZFS disks and your array. Are they single disk pools?

If I were to give you the ability to specify a list of ZFS pools, indicating wether or not they are part of the array or the cache, would this give you the flexibility to obtain the data in the format you require? I'm imagining additional entries to the cache/array response, one for each ZFS pool with the usual information.

I would essentially be parsing the output of  `zpool list` using the labels provided in the configuration to assign the parsed data to the array or the cache.

**Comment by NIronwolf** [comment]
I feel like all my snapshots are also adding to the jumble of numbers.

However, it seems that if a mountpoint gets listed in cache/array in the conf.yml and it happens to appear as the "Mountpoint" in the output of 'zfs list -p' you can then use the "Used" and "Avail" columns to sum up the total space and have the used/free values to return.

To clarify my setup, I have 13x 20TB drives in my array including 2 parity.  Disk 1 is my target for replication of my important appdata structure from nvme_cache. Disks 6, 10 and 11 just happened to be new and I made them zfs but they aren't getting special use.  Then there are 4x 2TB nvme drives that are in raidz1 working out to just under 6TB of space available (however it really seems to have always been shy of that figure).  I also have a SATA SSD and another USB HDD that are additional unraid cache pools, of which the SSD is a single disk zpool.

For whatever reason (probably because you can have multiple datasets in a pool) zfs seems to report the disk size as what's left available and zero usage.  I think the '**zfs** list' has the accurate data. The sizes in '**zpool** list' seem based on the total bytes of disks involved and over-report the total size when a disk is "lost" to parity.

```
zpool list -p
NAME                  SIZE           ALLOC            FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
disk1       19997367730176  13288895909888   6708471820288        -         -      8     66   1.00    ONLINE  -
disk10      19997367730176   9886393597952  10110974132224        -         -      0     49   1.00    ONLINE  -
disk11      19997367730176   9861560991744  10135806738432        -         -      0     49   1.00    ONLINE  -
disk6       19997367730176  14999326806016   4998040924160        -         -      0     75   1.00    ONLINE  -
nvme_cache   8177617731584   7333429780480    844187951104        -         -     46     89   1.00    ONLINE  -
ssd_cache    1992864825344     63720603648   1929144221696        -         -      3      3   1.00    ONLINE  -
```

[datasets.txt](https://github.com/NebN/unraid-simple-monitoring-api/files/14814135/datasets.txt) from **zfs** list -p

**Comment by NIronwolf** [comment]
Also, since you can have multiple unraid cache pools, do you think this is something that can be configured separately or do you plan to keep the result as the total off all the cache pools? (Probably a separate issue.)

**Comment by NebN** [comment]
I have done some testing. The solution you proposed seems to work, however it requires the container to run in privileged mode, this is something I'd like to avoid.

If I were to implement it this way, I would not want to make it the default, and only something individual users would need to opt in to.

In any case, if you feel comfortable with running the container as privileged I have made available a qa build that you can test out. If you wish to do so please follow the steps in[ this new section](https://github.com/NebN/unraid-simple-monitoring-api/tree/master?tab=readme-ov-file#qa) on the readme, while also selecting `Privileged: ON` when adding the new container.

**Comment by NebN** [comment]
> Also, since you can have multiple unraid cache pools, do you think this is something that can be configured separately or do you plan to keep the result as the total off all the cache pools? (Probably a separate issue.)

I am not sure exactly what you are asking, the response already provides separate data for each disk in the array, and for each cache listed, but yes please open a new issue if you wish to discuss this further.

**Comment by NIronwolf** [comment]
I setup the QA build and it's working just fine.  Thanks for taking the time to look in to it!  Now I can list the free GB on the array and each of the three cache pools on my dashboard.

I agree it would be nice to not have to run privileged. I often forget that's an issue since you always log in to unraid as root.

You're absolutely right about it already returning the data for each line in the cache section.  I think I'd been looking at adding up things for too long yesterday and was thinking my nvme were on separate lines for some reason.  No changes needed there.

**Comment by NebN** [comment]
That's great to hear. I'll figure out exactly how I want to deal with the higher privilege required, and then release it officially. In the meantime you can just keep using the QA build if you wish.

If you do, not update it as I'll very likely be pushing new experimental builds on that tag.

**Comment by NebN** [comment]
Released as part of [v0.3.0](https://github.com/NebN/unraid-simple-monitoring-api/releases/tag/v0.3.0)
You can now switch to the official image, just remember to set it to privileged if you wish to see the correct ZFS data.
